clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy

```diff
@@ -0,0 +1,206 @@
+    } else {
+      basicAuthenticate = null
+    }
+  }
+
+  /*
+   * Parsed according to http://www.ietf.org/rfc/rfc2617.txt
+   */
+  public AuthenticateDetails parseBearerAuthenticateHeader(String header) {
+    String bearerPrefix = "bearer "
+    String realmKey = "realm"
+    String serviceKey = "service"
+    String scopeKey = "scope"
+    AuthenticateDetails result = new AuthenticateDetails()
+
+    if (!bearerPrefix.equalsIgnoreCase(header.substring(0, bearerPrefix.length()))) {
+      throw new DockerRegistryAuthenticationException("Docker registry must support 'Bearer' authentication.")
+    } else {
+      header = header.substring(bearerPrefix.length())
+    }
+
+    // Each parameter has the form <token>=(<token>|<quoted-string>).
+    while (header.length() > 0) {
+      String key
+      String value
+
+      def keyEnd = header.indexOf("=")
+      if (keyEnd == -1) {
+        throw new DockerRegistryAuthenticationException("Www-Authenticate header terminated with junk: '$header'.")
+      }
+
+      key = header.substring(0, keyEnd)
+      header = header.substring(keyEnd + 1)
+      if (header.length() == 0) {
+        throw new DockerRegistryAuthenticationException("Www-Authenticate header unmatched parameter key: '$key'.")
+      }
+
+      // Parse a quoted string.
+      if (header[0] == '"') {
+        header = header.substring(1)
+
+        def valueEnd = header.indexOf('"')
+        if (valueEnd == -1) {
+          throw new DockerRegistryAuthenticationException('Www-Authenticate header has unterminated " (quotation mark).')
+        }
+
+        value = header.substring(0, valueEnd)
+        header = header.substring(valueEnd + 1)
+
+        if (header.length() != 0) {
+          if (header[0] != ",") {
+            throw new DockerRegistryAuthenticationException('Www-Authenticate header params must be separated by , (comma).')
+          }
+          header = header.substring(1)
+        }
+      } else { // Parse an unquoted token.
+        def valueEnd = header.indexOf(",")
+
+        // In the case of the last parameter, there will be no terminating ',' character.
+        if (valueEnd == -1) {
+          value = header
+          header = ""
+        } else {
+          value = header.substring(0, valueEnd)
+          header = header.substring(valueEnd + 1)
+        }
+      }
+
+      if (key.equalsIgnoreCase(realmKey)) {
+        def url = new URL(value)
+        result.realm = url.protocol + "://" + url.authority
+        result.path = url.path
+        if (result.path.length() > 0 && result.path[0] == "/") {
+          result.path = result.path.substring(1)
+        }
+      } else if (key.equalsIgnoreCase(serviceKey)) {
+        result.service = value
+      } else if (key.equalsIgnoreCase(scopeKey)) {
+        result.scope = value
+      }
+    }
+
+    if (!result.realm) {
+      throw new DockerRegistryAuthenticationException('Www-Authenticate header must provide "realm" parameter.')
```

Comment:
The exception message on line 54 uses single quotes, but you are using double quotes for these 3 exception messages.